src/Functions/toStartOfInterval.cpp

```diff
@@ -40,156 +44,227 @@ class FunctionToStartOfInterval : public IFunction
+            }
+            case ResultType::DateTime64:
+            {
+                UInt32 scale = 0;
+                if (interval_type->getKind() == IntervalKind::Nanosecond)
+                    scale = 9;
+                else if (interval_type->getKind() == IntervalKind::Microsecond)
+                    scale = 6;
+                else if (interval_type->getKind() == IntervalKind::Millisecond)
+                    scale = 3;
+
+                const size_t time_zone_arg_num = (arguments.size() == 2 || (arguments.size() == 3 && third_argument == ThirdArgument::IsTimezone)) ? 2 : 3;
+                return std::make_shared<DataTypeDateTime64>(scale, extractTimeZoneNameFromFunctionArguments(arguments, time_zone_arg_num, 0, false));
+            }
         }
 
-        UNREACHABLE();
+        std::unreachable();
     }
 
     ColumnPtr executeImpl(const ColumnsWithTypeAndName & arguments, const DataTypePtr & result_type, size_t /* input_rows_count */) const override
     {
         const auto & time_column = arguments[0];
         const auto & interval_column = arguments[1];
-        const auto & time_zone = extractTimeZoneFromFunctionArguments(arguments, 2, 0);
-        auto result_column = dispatchForTimeColumn(time_column, interval_column, result_type, time_zone);
+
+        ColumnWithTypeAndName origin_column;
+        const bool has_origin_arg = (arguments.size() == 3 && isDateOrDate32OrDateTimeOrDateTime64(arguments[2].type)) || arguments.size() == 4;
+        if (has_origin_arg)
+            origin_column = arguments[2];
+
+        const size_t time_zone_arg_num = (arguments.size() == 2 || (arguments.size() == 3 && isString(arguments[2].type))) ? 2 : 3;
+        const auto & time_zone = extractTimeZoneFromFunctionArguments(arguments, time_zone_arg_num, 0);
+
+        auto result_column = dispatchForTimeColumn(time_column, interval_column, origin_column, result_type, time_zone);
         return result_column;
     }
 
 private:
     ColumnPtr dispatchForTimeColumn(
-        const ColumnWithTypeAndName & time_column, const ColumnWithTypeAndName & interval_column, const DataTypePtr & result_type, const DateLUTImpl & time_zone) const
+        const ColumnWithTypeAndName & time_column, const ColumnWithTypeAndName & interval_column, const ColumnWithTypeAndName & origin_column, const DataTypePtr & result_type, const DateLUTImpl & time_zone) const
     {
-        const auto & from_datatype = *time_column.type.get();
+        const auto & time_column_type = *time_column.type.get();
+        const auto & time_column_col = *time_column.column.get();
 
-        if (isDateTime64(from_datatype))
+        if (isDateTime64(time_column_type))
         {
-            const auto * time_column_vec = checkAndGetColumn<ColumnDateTime64>(time_column.column.get());
-            auto scale = assert_cast<const DataTypeDateTime64 &>(from_datatype).getScale();
+            if (!isDateTime64(origin_column.type.get()) && origin_column.column != nullptr)
```

Comment:
The check would look better if the condition was reversed:

```cpp
if (origin_column.column != nullptr && !isDateTime64(origin_column.type.get()))
```

(reads: if the origin column is given and not of DateTime64 type, then ...)